### DISCUSSION TITLE
How does milvus handle recency ?

### DISCUSSION THREAD

**Discussion by jamesjohnson1025** [milvus-io/milvus#24695] [Q&A and General discussion] [answered]
I'm new to Milvus and currently exploring its functionalities. I have a question that I hope the Milvus community can help me with.

Let's consider a scenario where I have a document on Machine Learning that was initially drafted in 2014. I extracted the text from a PDF, converted it into embeddings, and stored them in Milvus. Later on, in 2023, I obtained an updated version of the document and followed the same process of extracting embeddings and storing them in Milvus.

Now, if I pose the same query, both documents contain the answer. However, the second document represents an updated version. My question is: When querying the system, will Milvus retrieve the answer from the first document (the initial version) or the second document (the latest version)?

**Comment by xiaofan-luan** [comment]
That's actually an interesting topic. 

So you are saying that you want a reranking policy respect the recency of your data?
Is it reasonable to retrieve your data with ts and score the data based on ts on data.

**Reply by jamesjohnson1025** [reply]
@xiaofan-luan  `Is it reasonable to retrieve your data with ts and score the data based on ts on data.`. How can we achieve this in milvus.

**Reply by jamesjohnson1025** [reply]
@xiaofan-luan  And one more question. Sorry to put this question in this discussion. How to add fields dynamically. There is no full fledged example. There is document inconsistency too. In this document https://milvus.io/docs/schema.md, there is no entry of dynamic field in schema but in this document https://milvus.io/api-reference/pymilvus/v2.2.x/Schema/CollectionSchema().md, there is an entry of dynamic field in schema. 

I tried both and it didn't work. So, i am asking if you have full fledged example to replicate it.

**Reply by jamesjohnson1025** [reply]
@xiaofan-luan Hi, Please let me know if it is not supported or still under development. Since it was part of documentation, i wanted to give a try. Thank you.

**Reply by jamesjohnson1025** [reply]
@xiaofan-luan Below is the example i tried. 

```
book_id = FieldSchema(
    name='boot_id',
    dtype=DataType.INT64,
    is_primary=True
)
embeddings = FieldSchema(
    name='embeddings',
    dtype=DataType.FLOAT_VECTOR,
    dim = 1536
)
community_id = FieldSchema(
    name='community_id',
    dtype=DataType.INT64,
)
message = FieldSchema(
    name='message',
    dtype=DataType.VARCHAR,
    max_length=65535
)

claps = FieldSchema(name="claps", dtype=DataType.INT64, is_dynamic=True)
remaining_time = FieldSchema(name="remaining_time",dtype=DataType.INT64,is_dynamic=True)


schema = CollectionSchema(
    fields=[book_id,embeddings,community_id,message],
    description='Johnson Test book search',
    auto_id=True,
    enable_dynamic_field=True
)

collection_name = 'jossy_Collection_book'
collection = Collection(
    name=collection_name, 
    schema=schema, 
    using='default', 
    shards_num=2,
    consistency_level="Strong"
)
```

When i try to enter the information

```
collection.insert([query_vector,[151],[new_sentence],[256],[2000]])
```
I get the below error 

```
message=The fields don't match with schema fields, expected: ['embeddings', 'community_id', 'message'], got ['', '', '', '', ''])>
```

**Reply by czs007** [reply]
@jamesjohnson1025 
sorry for any confusion caused by our previous documentation. To utilize the dynamic field feature, it suffices to simply pass enable_dynamic_field=True when creating the schema.
Currently, we only support row-level inserts for dynamic schema functionality.

Based on the example you previously provided, I have made some assumptions and have written the following code in hopes that it may be of assistance to you.
```
book_id = FieldSchema(
    name='boot_id',
    dtype=DataType.INT64,
    is_primary=True
)
embeddings = FieldSchema(
    name='embeddings',
    dtype=DataType.FLOAT_VECTOR,
    dim = 1536
)
community_id = FieldSchema(
    name='community_id',
    dtype=DataType.INT64,
)
message = FieldSchema(
    name='message',
    dtype=DataType.VARCHAR,
    max_length=65535
)


schema = CollectionSchema(
    fields=[book_id,embeddings,community_id,message],
    description='Johnson Test book search',
    auto_id=True,
    enable_dynamic_field=True
)

collection_name = 'jossy_Collection_book'
collection = Collection(
    name=collection_name, 
    schema=schema, 
    using='default', 
    shards_num=2,
    consistency_level="Strong"
)

rows = [
         { "book_id":1, "embeddings": vector_data1, "community_id": 1, "message": "the message1", "claps": 100, "remaining_time": 10,  },
         { "book_id":2, "embeddings": vector_data2, "community_id": 2, "message": "the message1", "claps": 101, "remaining_time": 11,  },
         { "book_id":3, "embeddings": vector_data3, "community_id": 3, "message": "the message1", "claps": 102, "remaining_time": 12,  },
         { "book_id":4, "embeddings": vector_data4, "community_id": 4, "message": "the message1", "claps": 103, "remaining_time": 13,  },

        ]

# insert a batch of rows
collection.insert(rows)

# insert one row
collection.insert({ "book_id":5, "embeddings": vector_data5, "community_id": 5, "message": "the message1", "claps": 104, "remaining_time": 14, })

```
`claps` and `remaining_time` are not included in the schema, but with enable_dynamic_field enabled, it is possible to specify these dynamic fields during both insertion and querying.

This is a functional example that you can refer to.  https://github.com/milvus-io/pymilvus/blob/2.2/examples/dynamic_field.py

**Reply by jamesjohnson1025** [reply]
Hi, I am getting `key error` when i ran the above code. My pymilvus version is == 2.2.0. I tried running the example you mentioned
``` https://github.com/milvus-io/pymilvus/blob/2.2/examples/dynamic_field.py``` but still it shows the same error. To pinpoint the line which causes error is 
``` fields = [FieldSchema("", infer_dtype_bydata(d[0])) for d in data]```. Its from schema.py and the function name is ```parse_fields_from_data```. Can you help me please

**Reply by jamesjohnson1025** [reply, marked as answer]
@czs007 Hi, its working for pymilvus 2.2.9. Thank you.

**Comment by yhmo** [comment]
@jamesjohnson1025 
I suppose the "claps" and "remaining_time" is dynamic values, the "remaining_time" is a timestamp value.

Then you don't need to create two FieldSchema for "claps" and "remaining_time", just turn on the switch "enable_dynamic_field".
The following script create a collection with 4 fields, enable_dynamic_field=tture. And insert two rows with extra dynamic values of  "claps" and "remaining_time".

"remaining_time=1850000000" is the old document.
"remaining_time=1860000000" is the new document.

Then I search the collection with the expression "remaining_time > 1850000000", this expression will filter out the old document. After search, only the new document item is in the result.

```
    connections.connect(host=HOST, port=PORT)
    print(utility.get_server_version())

    book_id = FieldSchema(
        name='boot_id',
        dtype=DataType.INT64,
        is_primary=True
    )
    embeddings = FieldSchema(
        name='embeddings',
        dtype=DataType.FLOAT_VECTOR,
        dim=1536
    )
    community_id = FieldSchema(
        name='community_id',
        dtype=DataType.INT64,
    )
    message = FieldSchema(
        name='message',
        dtype=DataType.VARCHAR,
        max_length=65535
    )

    schema = CollectionSchema(
        fields=[book_id, embeddings, community_id, message],
        description='Johnson Test book search',
        auto_id=True,
        enable_dynamic_field=True
    )

    collection_name = 'jossy_Collection_book'
    if utility.has_collection(collection_name):
        utility.drop_collection(collection_name)

    collection = Collection(
        name=collection_name,
        schema=schema,
        using='default',
        shards_num=2,
        consistency_level="Strong"
    )

    index = {
        "index_type": "IVF_FLAT",
        "metric_type": "L2",
        "params": {'nlist': 256},
    }
    collection.create_index("embeddings", index)
    collection.load()

    collection.insert({
        "embeddings": [random.random() for _ in range(1536)],
        "community_id": 151,
        "message": "old message",
        "claps": 123,
        "remaining_time": 1850000000,
    })

    collection.insert({
        "embeddings": [random.random() for _ in range(1536)],
        "community_id": 152,
        "message": "new message",
        "claps": 254,
        "remaining_time": 1860000000,
    })

    collection.flush()
    print("Row count:", collection.num_entities)

    query_vectors = [
        [random.random() for _ in range(1536)],
    ]
    search_param = {
        "expr": "remaining_time > 1850000000",
        "data": query_vectors,
        "anns_field": "embeddings",
        "param": {"metric_type": "L2", "params": {"nprobe": 32}},
        "output_fields": ["*"],
        "limit": 5}
    results = collection.search(**search_param)
    for i, result in enumerate(results):
        print("\nSearch result for {}th vector: ".format(i))
        for j, res in enumerate(result):
            print("Top {}: {}".format(j, res))
```

**Reply by jamesjohnson1025** [reply]
Any guess what will be the type of value of timestamp.

**Reply by yhmo** [reply]
To make the "remaining_time" comparable, it is better to set it as integer type. Because integer value is easy to do filtering: https://milvus.io/docs/boolean.md

**Reply by jamesjohnson1025** [reply]
Thank you. One more last question. Is there any `sorting function` thats part of milvus where you can sort the resultant records apart from the `sort function` provided by Python library

**Reply by yhmo** [reply]
> Thank you. One more last question. Is there any `sorting function` thats part of milvus where you can sort the resultant records apart from the `sort function` provided by Python library

I don't understand. The results returned by search() are already sorted by distance. Why do you want to sort them again?
